### ISSUE TITLE
fine-tuning mBART for monolingual summarization

### ISSUE THREAD

**Issue by blackcat84** [pytorch/fairseq#1924]
I was wondering whether it'd be possible to fine-tune mBART for summarization on one of the languages used during pre-training. Judging from the way mBART (cc25) was pre-trained (language to same language instead of cross-language) I suppose it's doable?
My idea is to fine-tune by simply taking the text as input to the encoder and the summary as input/output of the decoder.
Do you have any tips or in general something I should be worried about?

Thanks!

**Comment by didge-rydu** [comment]
+1

It seems like the code doesn't allow for this. The input and output language need to be different, since they are used to differentiate input and output data in the code.  This leaves me wondering: Is there any reason this shouldn't be done using the normal translation script?
Anyways, I got it to work by changing lines 46, 66 and 77 of the following file
https://github.com/pytorch/fairseq/blob/master/fairseq/tasks/translation.py
to allow for input and output to have the same language in their filename.

**Comment by blackcat84** [comment]
Are you sure the languages need to be different? I was able to run the code by simply naming source and target language as src and tgt (in argument parsing), but having in reality the same language in input and output

**Comment by didge-rydu** [comment]
Ok that is interesting. So you just ran
train.py --source-lang src --target-lang tgt ..
I initially thought that these command line arguments are used for more things than just loading the dataset, but maybe this is in fact the intended way to use them. For mBART, the dictionary is the same anyways for all languages.

**Comment by blackcat84** [comment]
Yes I think that should be fine, this is similar to what is suggested for the monolingual BART here
https://github.com/pytorch/fairseq/blob/master/examples/bart/README.summarization.md
expect I had to change the task from translation to translation from pretrained

**Comment by gusdnd852** [comment]
I have a question for fine-tuning monolingual mbart: when I'm fine-tuning mbart, should I add the language code directly? If it is automatically added, which stage does the language code add? Also, is there a way to set up only one language to be added?

**Comment by gusdnd852** [comment]
I can run the train cli code with ` --source-lang src --target-lang tgt`. but in this case, how language id works? (like ko_KR, en_XX, ro_RO, etc...) for mBART training, I should add the languaged code as eos token.

**Comment by blackcat84** [comment]
I'm not sure what you mean by 

> I should add the languaged code as eos token.

but as long as you are using a language that was included during pre-training I don't think you need to worry, if I remember correctly the vocabulary is the same for all languages and the language id in your case would simply be `src` and `tgt`

**Comment by gusdnd852** [comment]
When mBART is pre-learning, it is learned that the decoder must receive ko_KR as bos to output Korean, and the decoder must receive ar_AR as bos to output in Arabic.

However, if we set something like tgt or src to language, <unk> is entered as the encoder's eos and decoder's bos (see code for this). Returns <unk> if there is no such language in the dictionary dict.index(language) -> <unk> -> eos 

This is not a problem because BART, (not mBART), uses tokens such as <s> and </s> as bos and eos, but in mBART, It is a problem because it uses language token like ko_KR, ar_AR as bos and eos.

**Comment by blackcat84** [comment]
Right, I see what you mean.

Yes using `src` and `tgt` and therefore `unk` is far from ideal. I'm wondering how much this hurts the performance. Considering also that the <LID> token is used to stop decoding, I think a simple dirty trick is to define a maximum number of tokens and stop decoding when that threshold is reached.

Another solution would be using the same <LID> for both source and target, e.g. [ko_KR]. If I remember correctly there's a check that doesn't allow this, but it can be easily commented out. However I'm not sure if this choice influences other sections or if in general constitutes a problem.

**Comment by gusdnd852** [comment]
The biggest problem is the decoder's bos. [ko_KR] must be entered as bos for mBART to properly output Korean. (Because the model learned that way) But when an unk token is put into the decoder's bos, the output is very weird. This has a profound effect on performance. When I used the same LID (ko_KR) for both eos and bos using Huggingface's transformers, the NLL loss reached 0.9 for one epoch, but the NLL loss stayed at about 3 to 4 when I learned with Fairseq.

**Comment by clang88** [comment]
> The biggest problem is the decoder's bos. [ko_KR] must be entered as initial token (= bos token) for mBART Decoders to properly output Korean. (Because the model learned that way) But when an `<unk>` token is put into the decoder's bos, the output is very weird. This has a profound effect on performance. When I used the same LID (ko_KR) for both eos and bos using Huggingface's transformers, the NLL loss reached 0.9 for one epoch and output's quaility is very good. but the NLL loss stayed at about 3 to 4 and output is very weird when I used src, tgt as language id with Fairseq
> 
> Originally, I was implementing it using a Huggingface transformers, but I have to solve this problem because I am migrating the work to Fairseq in the project that I am currently working with my team members.

Thanks for the great investigative effort!
Did you make any progress since then? I am currently trying to use mBART on a variety of downstream tasks (always source lang -> source lang) and would like to do multilingual training. I have already achieved good results just using "src" as source language and "label" as target language. I guess this is because in my downstream tasks the decoder never has to generate full sentences, but I would still like to try it with the right eos and bos tags applied. 

I've noticed the <unk> eos tag when generating on the test set, so I guess the Encoder is not performing at its best, but I've also noticed that the code states "prepend_bos" as "False", which seems to imply at no bos tag is actually automatically added to the data? In fact, I can't see any <unk> dag prepended to the "gold" translation when generating on my test set.

I guess without drastically changing the code, the best I could do is manually append/prepend the tags to my data and set append_source_id=False,  right? Either way, I will try this and report back.

**Comment by dimtsi** [comment]
Has there been any progress on this? I am also trying to train mBART on a summarization task in one of the languages it was pretrained on.

**Comment by jonathanmutal** [comment]
Has there been any progress on this? I am also trying to train mBART on a summarization task in one of the languages it was pretrained on.

**Comment by KpKqwq** [comment]
I am training mbart on spanish. I changed the origianl code as follows:
1).https://github.com/facebookresearch/fairseq/blob/ad0e69cd99e1ff884041fbd8467d1404bd09847a/fairseq/tasks/translation.py#L138 add src=tgt="es_XX"
2)https://github.com/facebookresearch/fairseq/blob/ad0e69cd99e1ff884041fbd8467d1404bd09847a/fairseq/tasks/translation_from_pretrained_bart.py#L92 change the self.args.target_lang to "es_XX"  in fucntion build_generator"
3)https://github.com/facebookresearch/fairseq/blob/ad0e69cd99e1ff884041fbd8467d1404bd09847a/fairseq/tasks/translation_from_pretrained_bart.py#L119 change the self.args.target_lang to "es_XX" in fucntion build_dataset_for_inference
